### ISSUE TITLE
strange behaviour with Anima in Forge NEO for some days

### ISSUE THREAD

**Issue by Xeltosh** [Haoming02/sd-webui-forge-classic#1112]
Hello again,

I am not sure if this is a bug or not, as it is something that happens without an error message. Last time i used the affected prompts and where they worked was May 2nd. So maybe one change since then made a problem in the interpretation of prompts.

It is best described as: everything looks less detailed and more crude/unrefined. sadly I can't really give a more detailed explanation right now.

I am a fan of using more tokens and many of my prompts don't work anymore. First i thought it is a length problem but after testing around with prompts that work has disproven this in my opinion. A pic with 463 tokens works as intended and a different picture with a different outfit (but the rest is the same) with 456 tokens doesn't work anymore. Tried reading over every prompt but i can't really make out what's wrong. 

Decided to gen one picture with the problem and try different stuff i did since May 2nd to see, what causes the issue.

- I first deactivated most of the support stuff (--flash --sage and so on), no change
- deleted my venv and let it rebuild again -> problem persists
- changing/removing some prompts didn't work either
- last resort: trying to roll back every commit you made since May 2nd
-> found the culprit: the commit 28e3947 from May 5th is the last commit that fully works with my prompts. 

Something you changed with your commit cd2490a from May 7th changed how my prompts are interpreted in the background and broke stuff for me. I can't say if it is prompt/word related or not, but i have found out, that prompts, that want a "specific amount of detail" e.g. intricate gold chains or elaborate cloth designs are mostly breaking stuff.

I have no clue if i am the only one affected, but i hope you can fix it. I always like to try stuff out and help when i have the free time, so feel free to look me up in Discord. I am also a moderator on Mobians.ai and helped the admin there to develop his model back in the day with SD1.5 and SDXL and also helped crody with some stuff regarding his Sonic-Illoustrious model, so i have some experience in troubleshooting stuff.

**Comment by nounainoise** [comment]
When I run Anima on commits newer than May 1st or so, I get the message `Anima Unexpected: ['pos_embedder.dim_spatial_range', 'pos_embedder.dim_temporal_range', 'pos_embedder.seq']' and poor generations as you've pointed out.

**Comment by Xeltosh** [comment]
> When I run Anima on commits newer than May 1st or so, I get the message `Anima Unexpected: ['pos_embedder.dim_spatial_range', 'pos_embedder.dim_temporal_range', 'pos_embedder.seq']' and poor generations as you've pointed out.

could you gen one poor picture and try to rollback every commit and gen the same picture until they stop being poor? i found the commit that makes the problem for me like this. maybe we can pinpoint something. if it doesn't stop, the problem should be on your side.

**Comment by nounainoise** [comment]
> > When I run Anima on commits newer than May 1st or so, I get the message `Anima Unexpected: ['pos_embedder.dim_spatial_range', 'pos_embedder.dim_temporal_range', 'pos_embedder.seq']' and poor generations as you've pointed out.
> 
> could you gen one poor picture and try to rollback every commit and gen the same picture until they stop being poor? i found the commit that makes the problem for me like this. maybe we can pinpoint something. if it doesn't stop, the problem should be on your side.

I should've really cleared my venv each time, but I tried four different commits: ae9ce94e880fb963754cd2c4cd4837eab6e52cc6 (latest), 28e39472a1ec86a2b8f2e626c96ed272ff3dd5f4 (the one you pointed out), cd2490a8d285000c33cab8471a4335fc4e911c53 (your "good" one), ae03079572e1445f353ec43d0f2bc3a8c393be0c (my "good" one). With the same prompt, same settings and seed only the latest commit was visually distinct. The other three generations were exactly the same. I wouldn't rule out a problem on my end. I noticed the issue a few days ago and turned off sage attention and things did improve slightly.

**Comment by Haoming02** [comment]
Can't repro...

<br>

<p align="center">
<img src="https://github.com/user-attachments/assets/1ae5a80c-4748-4306-a9a8-9569df3c7c6d" />
<br>
Generated on <b>Apr. 29</b>
</p>

<p align="center">
<img src="https://github.com/user-attachments/assets/47b68240-8b89-447f-b85e-fb6d6fd4b421" />
<br>
Generated on <b>May. 10</b>
</p>

<hr>

```
masterpiece, best quality, very aesthetic, absurdres, arima kana, oshi no ko, @[hirayama kanna|mika pikazo|sy4|ogipote], 1girl, photo background, real world location, red eyes, red hair, inverted bob, bob cut, open mouth, beret, smile, blush, white gloves, solo, night sky, happy, looking at viewer, upper body, outdoors, medium hair.
Arima Kana from Oshi no Ko featuring shoulder-length red hair in an inverted bob cut with bangs and loose strands flowing in the wind. She offers a light smile while looking at the viewer with her striking red eyes, holding a canned coffee up near her chin with both hands. She wears a winter ensemble defined by a thick fur collar and fur-trimmed sleeves, with a large white bow visible on her back. The background is a realistic night scene set in Kobe, Hyogo Prefecture, showing waterfront buildings and a blue firework display over the water, surrounded by falling white particles that resemble snow or light sparkles.
Negative prompt: worst quality, low quality, bad anatomy, artistic error
Steps: 30, Sampler: Res Multistep, Schedule type: SGM Uniform, CFG scale: 4, Shift: 3, Seed: 39457474, Size: 1536x1024, Model hash: ed5e5bcdfa, Model: kirazuriAnima_v20AnimaPreview3, Module 1: qwen_3_06b, Module 2: qwen_image_vae, Clip skip: 2, RNG: CPU, Version: neo
```

**Comment by nounainoise** [comment]
I tweaked things to match my usual workflow, but with ADetailer off to rule that out as a factor (but I have been using abzaloff's version for Neo)
(244 tokens)
Steps: 30, Sampler: Res Multistep, Schedule type: Beta, CFG scale: 6, Shift: 3, Seed: 3492219790, Size: 1216x832, Model hash: ebefe4ebf0, Model: hikariAnima_041, Module 1: qwen_image_vae, Module 2: qwen_3_06b_base, Diffusion in Low Bits: Automatic (fp16 LoRA), Denoising strength: 0.35, Clip skip: 2, RNG: CPU, Hires Module 1: Use same choices, Hires upscale: 1.25, Rescale CFG: 0.3, Beta schedule alpha: 0.6, Beta schedule beta: 0.6, Hires Shift: 6, Hires schedule type: SGM Uniform, Hires upscaler: 4x_foolhardy_Remacri, Hires steps: 32, Hires CFG Scale: 6, 

ae9ce94e880fb963754cd2c4cd4837eab6e52cc6
<img width="1520" height="1040" alt="Image" src="https://github.com/user-attachments/assets/4a3da559-ead2-488b-8523-a3e304bed3ad" />

28e39472a1ec86a2b8f2e626c96ed272ff3dd5f4
<img width="1520" height="1040" alt="Image" src="https://github.com/user-attachments/assets/21fb98f7-58df-4969-9daa-74e68fde94a6" />

The difference is very slight here, but with my usual prompts that are 600+ tokens (that I can't post here, they would break your rules), the difference is drastic.

**Comment by Xeltosh** [comment]
@nounainoise this may sound dumb but what rules do you mean? I don't want to run into problems trying to show my problem 😅

**Comment by Xeltosh** [comment]
Wrong button pressed on my mobile, didn't want to close this....

what i found out after some short testing to properly reproduce this error is, that it probably has something to do with prompt length and specific prompts.

i interchanged some outfits, some backgrounds, poses and some characters. sometimes one background and outfit combo on one char worked, on another it didn't. will need to test some more to get something proper.

It is just strange, that all combinations work in the 28e3947 commit, but make this error after every following commit....

**Comment by plularias** [comment]
I can confirm this is very likely related to commit [cd2490a](https://github.com/Haoming02/sd-webui-forge-classic/commit/cd2490a8d285000c33cab8471a4335fc4e911c53) and specifically to the prompt padding change in `backend/text_processing/anima_engine.py`.
On my side, reverting only that padding logic fixes the problem

The relevant change was essentially this:

```python
# old
if cross_attn.shape[1] < 512:
    cross_attn = torch.nn.functional.pad(cross_attn, (0, 0, 0, 512 - cross_attn.shape[1]))

# new
dim = math.ceil(cross_attn.shape[1] / 512) * 512
cross_attn = torch.nn.functional.pad(cross_attn, (0, 0, 0, dim - cross_attn.shape[1]))
```

So before, the Anima conditioning sequence was only padded to a **minimum** length of `512`; if `cross_attn.shape[1]` was already `>= 512`, it was left unchanged. After `cd2490a`, that sequence length is rounded up to a multiple of `512` instead.

For example, a sequence length of `520` used to stay `520`, but after `cd2490a` it gets padded to `1024`.

My guess is that this changes Anima prompt conditioning in a harmful way, because those extra padded rows are not being ignored properly in the downstream cross-attention path. In practice, that means prompts above 512 tokens can get a lot of extra "empty" conditioning rows, which seems to dilute the real prompt information instead of preserving it.

That matches the symptoms very well:

- prompts still "work", but they become less precise
- fine details / intricate design elements are weakened
- the result looks more crude / less refined
- reverting this one change restores the old behavior

**Comment by Haoming02** [comment]
Reverted that commit~

**Comment by Xeltosh** [comment]
> Reverted that commit~

yep, can confirm problem is gone.
